rasterio/_drivers.pyx

```diff
@@ -69,20 +70,77 @@ def driver_count():
     return GDALGetDriverCount() + OGRGetDriverCount()
 
 
-cdef class GDALEnv(object):
+cdef class ConfigEnv(object):
 
-    cdef object is_chef
     cdef public object options
+    cdef public object prev_options
+
+    def __init__(self, **options):
+        self.options = options.copy()
+        self.prev_options = {}
+
+    def enter_config_options(self):
+        """Set GDAL config options."""
+        cdef const char *key_c
+        cdef const char *val_c
+
+        for key, val in self.options.items():
+            key_b = key.upper().encode('utf-8')
+            key_c = key_b
+
+            # Save current value of that key.
+            val_c = CPLGetConfigOption(key_c, NULL)
+            if val_c != NULL:
+                val_b = val_c
+                self.prev_options[key_b] = val_b
+
+            if isinstance(val, string_types):
+                val_b = val.encode('utf-8')
+            else:
+                val_b = ('ON' if val else 'OFF').encode('utf-8')
+            val_c = val_b
+            CPLSetConfigOption(key_c, val_c)
+            log.debug("Option %s=%s", key, CPLGetConfigOption(key_c, NULL))
+
+    def exit_config_options(self):
+        """Clear GDAL config options."""
+        cdef const char *key_c
+        cdef const char *val_c
+
+        for key in self.options:
+            key_b = key.upper().encode('utf-8')
+            key_c = key_b
+            if key_b in self.prev_options:
+                val_b = self.prev_options[key_b]
+                key_c = key_b; val_c = val_b
+                CPLSetConfigOption(key_c, val_c)
+            else:
+                CPLSetConfigOption(key_c, NULL)
```

Comment:
@sgillies  I wonder if using `CPLSetConfigOption` instead of `CPLSetThreadLocalConfigOption` might solve #599